render.go

```diff
@@ -0,0 +1,87 @@
+package render
+
+import (
+	"image/color"
+	"image/draw"
+	"math"
+
+	"github.com/benoitkugler/textlayout/fonts"
+	"github.com/go-text/typesetting/shaping"
+	"github.com/srwiley/rasterx"
+	"golang.org/x/image/math/fixed"
+)
+
+type Renderer struct {
+	FontSize, PixScale float32
+	Color              color.Color
+}
+
+func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
+	sh := &shaping.HarfbuzzShaper{}
```

Comment:
There are significant efficiency gains from reusing a shaper instead of re-creating one each time you rasterize. Perhaps this should be a field on Renderer, assuming that Renderers are meant to be reused? Similarly, if you do any line wrapping, reusing the wrapper is much more efficient.

Reply:
Well, I tried it, and it seems like the raster process totally dominates, so it's a very minor efficiency gain. If you'd like to experiment, here is what I did:

```diff
commit dab5cca71a7e29d5dabd2efba0fb6ebf7bd24aa0
Author: Chris Waldon <christopher.waldon.dev@gmail.com>
Date:   Wed Feb 1 16:47:39 2023 -0500

    render: reuse shaper
    
    Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

diff --git a/render.go b/render.go
index 8f12dca..d6aae68 100644
--- a/render.go
+++ b/render.go
@@ -14,10 +14,10 @@ import (
 type Renderer struct {
 	FontSize, PixScale float32
 	Color              color.Color
+	shaper             shaping.HarfbuzzShaper
 }
 
 func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
-	sh := &shaping.HarfbuzzShaper{}
 	in := shaping.Input{
 		Text:     []rune(str),
 		RunStart: 0,
@@ -25,12 +25,11 @@ func (r *Renderer) DrawString(str string, img draw.Image, face fonts.Face) int {
 		Face:     face,
 		Size:     fixed.I(int(r.FontSize * r.PixScale)),
 	}
-	out := sh.Shape(in)
+	out := r.shaper.Shape(in)
 	return r.DrawShapedRunAt(out, img, 0, out.LineBounds.Ascent.Ceil())
 }
 
 func (r *Renderer) DrawStringAt(str string, img draw.Image, x, y int, face fonts.Face) int {
-	sh := &shaping.HarfbuzzShaper{}
 	in := shaping.Input{
 		Text:     []rune(str),
 		RunStart: 0,
@@ -38,7 +37,7 @@ func (r *Renderer) DrawStringAt(str string, img draw.Image, x, y int, face fonts
 		Face:     face,
 		Size:     fixed.I(int(r.FontSize * r.PixScale)),
 	}
-	return r.DrawShapedRunAt(sh.Shape(in), img, x, y)
+	return r.DrawShapedRunAt(r.shaper.Shape(in), img, x, y)
 }
 
 func (r *Renderer) DrawShapedRunAt(run shaping.Output, img draw.Image, startX, startY int) int {

commit 08359a8ae889a3240b92dddaaea95e0e4eff2f8f
Author: Chris Waldon <christopher.waldon.dev@gmail.com>
Date:   Wed Feb 1 16:33:38 2023 -0500

    render: add simple benchmark
    
    Signed-off-by: Chris Waldon <christopher.waldon.dev@gmail.com>

diff --git a/render_test.go b/render_test.go
index 2edcc0b..01a005b 100644
--- a/render_test.go
+++ b/render_test.go
@@ -41,3 +41,22 @@ func Test_Render(t *testing.T) {
 	png.Encode(w, img)
 	w.Close()
 }
+
+func Benchmark_Render(b *testing.B) {
+	img := image.NewNRGBA(image.Rect(0, 0, 350, 180))
+	draw.Draw(img, img.Bounds(), &image.Uniform{color.White}, image.Point{}, draw.Src)
+
+	data, _ := os.Open("testdata/NotoSans-Regular.ttf")
+	f, _ := font.ParseTTF(data)
+
+	r := &render.Renderer{
+		FontSize: 24,
+		PixScale: float32(2),
+		Color:    color.Black,
+	}
+	str := "Hello! ± सभमन"
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		r.DrawString(str, img, f)
+	}
+}
```